src/comm.jl

```diff
@@ -1,264 +1,1887 @@
+         offset = pmsg * maxPlaneComm + emsg * maxEdgeComm
+         req = irecv!(fromProc, offset, recvCount)
+         domain.recvRequest[pmsg+emsg+1] = req
          emsg += 1
       end
 
-      if (colMin && planeMin && doRecv)
-          fromRank = myRank - domain.m_tp*domain.m_tp - 1
-        start = pmsg * maxPlaneComm + emsg * maxEdgeComm
-      buf = @view domain.commDataRecv[start:start + dy *xferFields]
-      domain.recvRequest[pmsg+emsg] = MPI.Irecv!(buf, fromRank, msgType, comm)
+      if colMin && planeMin && doRecv
+         fromProc = myRank - domain.tp^2 - 1
+         recvCount = dy * xferFields
+         offset = pmsg * maxPlaneComm + emsg * maxEdgeComm
+         req = irecv!(fromProc, offset, recvCount)
+         domain.recvRequest[pmsg+emsg+1] = req
          emsg += 1
       end
 
-      if (rowMax && colMax)
-          fromRank = myRank + domain.m_tp + 1
-        start = pmsg * maxPlaneComm + emsg * maxEdgeComm
-      buf = @view domain.commDataRecv[start:start + dz *xferFields]
-      domain.recvRequest[pmsg+emsg] = MPI.Irecv!(buf, fromRank, msgType, comm)
+      if rowMax && colMax
+         fromProc = myRank + domain.tp + 1
+         recvCount = dz * xferFields
+         offset = pmsg * maxPlaneComm + emsg * maxEdgeComm
+         req = irecv!(fromProc, offset, recvCount)
+         domain.recvRequest[pmsg+emsg+1] = req
          emsg += 1
       end
 
-      if (rowMax && planeMax)
-          fromRank = myRank + domain.m_tp*domain.m_tp + domain.m_tp
-        start = pmsg * maxPlaneComm + emsg * maxEdgeComm
-      buf = @view domain.commDataRecv[start:start + dx *xferFields]
-      domain.recvRequest[pmsg+emsg] = MPI.Irecv!(buf, fromRank, msgType, comm)
+      if rowMax && planeMax
+         fromProc = myRank + domain.tp^2 + domain.tp
+         recvCount = dx * xferFields
+         offset = pmsg * maxPlaneComm + emsg * maxEdgeComm
+         req = irecv!(fromProc, offset, recvCount)
+         domain.recvRequest[pmsg+emsg+1] = req
          emsg += 1
       end
 
-      if (colMax && planeMax)
-          fromRank = myRank + domain.m_tp*domain.m_tp + 1
-        start = pmsg * maxPlaneComm + emsg * maxEdgeComm
-      buf = @view domain.commDataRecv[start:start + dy *xferFields]
-      domain.recvRequest[pmsg+emsg] = MPI.Irecv!(buf, fromRank, msgType, comm)
+      if colMin && planeMax
+         fromProc = myRank + domain.tp^2 + 1
+         recvCount = dy * xferFields
+         offset = pmsg * maxPlaneComm + emsg * maxEdgeComm
+         req = irecv!(fromProc, offset, recvCount)
+         domain.recvRequest[pmsg+emsg+1] = req
          emsg += 1
       end
 
-      if (rowMax && colMin)
-          fromRank = myRank + domain.m_tp - 1
-        start = pmsg * maxPlaneComm + emsg * maxEdgeComm
-      buf = @view domain.commDataRecv[start:start + dz *xferFields]
-      domain.recvRequest[pmsg+emsg] = MPI.Irecv!(buf, fromRank, msgType, comm)
+      if rowMax && colMin
+         fromProc = myRank + domain.tp + 1
```

Comment:
Minus one for fromProc: https://github.com/wsmoses/Enzyme-MPI/blob/d211d3522628628cc3990a2470639b3159104a37/lulesh-comm.cc#L219